odoo/tools/translate.py

```diff
@@ -886,107 +914,152 @@ def push_translation(module, type, name, id, source, comments=None):
         if not sanitized_term or len(sanitized_term) <= 1:
             return
 
-        tnx = (module, source, name, id, type, tuple(comments or ()))
-        to_translate.add(tnx)
+        self._to_translate.append((module, source, name, res_id, ttype, tuple(comments or ())))
+
+    # TODO remove me, I am batched !
+    def _get_translatable_record(self, model, res_id):
+        if model not in self.env:
+            _logger.error("Unable to find object %r", model)
+            return False
+
+        original = record = self.env[model].browse(res_id)
+        if not record.exists():
+            _logger.warning("Unable to find object %r with id %d", model, res_id)
+            return False
 
-    def translatable_model(record):
         if not record._translate:
             return False
 
         if record._name == 'ir.model.fields.selection':
             record = record.field_id
         if record._name == 'ir.model.fields':
             field_name = record.name
-            field_model = env.get(record.model)
+            field_model = self.env.get(record.model)
             if (field_model is None or not field_model._translate or
                     field_name not in field_model._fields):
                 return False
 
-        return True
+        return original
 
-    query = 'SELECT min(name), model, res_id, module FROM ir_model_data'
+    def _get_translatable_records(self, records):
+        """ Filter the records that are translatable
 
-    if 'all_installed' in modules:
-        query += ' WHERE module IN ( SELECT name FROM ir_module_module WHERE state = \'installed\') '
+        A record is considered as untranslatable if:
+        - it does not exist
+        - the model is flagged with _translate=False
+        - it is a field of a model flagged with _translate=False
+        - it is a selection of a field of a model flagged with _translate=False
 
-    if 'all' not in modules:
-        query += ' WHERE module IN %s'
-        query_param = (tuple(modules),)
-    else:
-        query += ' WHERE module != %s'
-        query_param = ('__export__',)
+        :param records: a list of namedtuple belonging to the same model
+        """
+        model = next(iter(records)).model
+        if model not in self.env:
+            _logger.error("Unable to find object %r", model)
+            return False
 
-    query += ' GROUP BY model, res_id, module ORDER BY module, model, min(name)'
+        if not self.env[model]._translate:
+            return False
```

Comment:
The method should always return a recordset, possibly empty.
This would avoid the necessary
```python
if not records:
    continue
```
in the caller's loop below.